engine/src/main/java/io/camunda/zeebe/engine/state/appliers/ProcessInstanceModifiedEventApplier.java

```diff
@@ -7,31 +7,74 @@
  */
 package io.camunda.zeebe.engine.state.appliers;
 
+import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableFlowNode;
 import io.camunda.zeebe.engine.state.TypedEventApplier;
 import io.camunda.zeebe.engine.state.instance.ElementInstance;
 import io.camunda.zeebe.engine.state.mutable.MutableElementInstanceState;
+import io.camunda.zeebe.engine.state.mutable.MutableProcessState;
 import io.camunda.zeebe.protocol.impl.record.value.processinstance.ProcessInstanceModificationRecord;
 import io.camunda.zeebe.protocol.record.intent.ProcessInstanceModificationIntent;
+import io.camunda.zeebe.protocol.record.value.BpmnElementType;
+import java.util.Collections;
 
 final class ProcessInstanceModifiedEventApplier
     implements TypedEventApplier<
         ProcessInstanceModificationIntent, ProcessInstanceModificationRecord> {
 
   private final MutableElementInstanceState elementInstanceState;
+  private final MutableProcessState processState;
 
   public ProcessInstanceModifiedEventApplier(
-      final MutableElementInstanceState elementInstanceState) {
+      final MutableElementInstanceState elementInstanceState,
+      final MutableProcessState processState) {
     this.elementInstanceState = elementInstanceState;
+    this.processState = processState;
   }
 
   @Override
   public void applyState(final long key, final ProcessInstanceModificationRecord value) {
-    value.getActivatedElementInstanceKeys().stream()
+    if (value.hasActivateInstructions()) {
+      clearInterruptedState(value);
+      incrementNumberOfTakenSequenceFlows(value);
+    }
+  }
+
+  private void clearInterruptedState(final ProcessInstanceModificationRecord value) {
+    value.getAncestorScopeKeys().stream()
         .map(elementInstanceState::getInstance)
         .filter(ElementInstance::isInterrupted)
         .forEach(
             instance ->
                 elementInstanceState.updateInstance(
                     instance.getKey(), ElementInstance::clearInterruptedState));
   }
+
+  private void incrementNumberOfTakenSequenceFlows(final ProcessInstanceModificationRecord value) {
+    final ElementInstance processInstance =
+        elementInstanceState.getInstance(value.getProcessInstanceKey());
+    final var process =
+        processState
+            .getProcessByKey(processInstance.getValue().getProcessDefinitionKey())
+            .getProcess();
+    value
+        .getActivateInstructions()
+        .forEach(
+            instruction -> {
+              final var element = process.getElementById(instruction.getElementId());
```

Comment:
:wrench: We could avoid the casting of the element below by using the following method.

```suggestion
              final var element =
                  process.getElementById(instruction.getElementId(), ExecutableFlowNode.class);
```